### PR TITLE
fix(no-leaked-conditional-rendering): same test and consequent in conditional expression

### PR DESCRIPTION
**Pull request by hyoban**
Not quite sure if I'm right

### CODE REVIEW

File: packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.spec.ts [hunk 1]
```diff
@@ -299,6 +299,21 @@ ruleTester.run(RULE_NAME, rule, {
...
       }
     `,
+    dedent`
+      const someCondition1 = 0;
+      const SomeComponent = () => <div />;
+
+      const App = () => {
+        return (
+          <>
+            {someCondition
```
Comment (Rel1cx) [inline packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.spec.ts:311]: The test case above is reporting an error because
The identifier `someCondition` (but `someCondition1` exists, make sure it's spelled correctly) is not in the code of this test case, so it is inferred to be of type `any`, which is a non-safe (as consequent node) type, so it is expected behavior.

**Review by Rel1cx — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.spec.ts (hunk 1 above)
Comment (hyoban) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.spec.ts:311]: Sorry, I wrote the wrong test, updated.

Comment (Rel1cx) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.spec.ts:311]: It looks like the changes made to no-leaked-conditional-rendering.ts are not needed. I'll add more tests soon to find edge cases for that rule.

File: packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts [hunk 2]
```diff
@@ -206,7 +206,14 @@ export default createRule<[], MessageID>({
     function isValidConditionalExpression(
       node: TSESTree.ConditionalExpression,
     ): boolean {
-      const { alternate, consequent } = node;
+      const { alternate, consequent, test } = node;
```
Comment (Rel1cx) [inline packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: In this scenario we don't care about the value of the test node at all, whatever it is will never be inserted into JSX and rendered by react (only consequent and alternate will).

**Review by Rel1cx — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts (hunk 2 above)
Comment (hyoban) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: What I'm thinking about is that I might just want to render this number when it's non-null or non-0.

```
x ? x : fallback
```

Comment (Rel1cx) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: > What I'm thinking about is that I might just want to render this number when it's non-null or non-0.
> 
> ```
> x ? x : fallback
> ```

That will not work, in the updated test case you have annotated the type to someCondition?: number | undefined, which include falsy number 0.

Only when it's type is non-zero number literal types like 2, 3, 5... make sense e.g. someCondition?: 5 | undefined, the expression in test node did check for that but the code added for this case has not check the type of the identifier before return true (not checked non-number type which we can't tell if it's safe).

Comment (Rel1cx) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: > > What I'm thinking about is that I might just want to render this number when it's non-null or non-0.
> > ```
> > x ? x : fallback
> > ```
> 
> That will not work, in the updated test case you have annotated the type to someCondition?: number | undefined, which include falsy number 0.
> 
> Only when it's type is non-zero number literal types like 2, 3, 5... make sense e.g. someCondition?: 5 | undefined, the expression in `test` node did check for that but the code added for this case has not check the type of the identifier before return true (not checked non-number type which we can't tell if it's safe).

Some information was incorrect and has been updated.

Comment (Rel1cx) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: I apologize for the inaccuracy of the information I provided above.
I looked at it, We can indeed check if the `consequent` is the same as `test`!

Add

```ts
import { isNodeEqual } from "@eslint-react/ast";

const allowGuardedTypes = [
  "boolean",
  "string",
  "number",
  "nullish",

  "truthy boolean",
  "truthy number",
  "truthy string",
] as const;
```

And then change the `isValidConditionalExpression` function into this:

```ts
function isValidConditionalExpression(
  node: TSESTree.ConditionalExpression,
): boolean {
  const { alternate, consequent, test } = node;

  const isConsequentGuarded = isNodeEqual(consequent, test);
  const testType = getConstrainedTypeAtLocation(services, test);
  const types = inspectVariantTypes([testType]);

  if (isConsequentGuarded && types.every(type => allowGuardedTypes.includes(type as never))) {
    return true;
  }

  return isValidExpression(alternate) && isValidExpression(consequent);
}
```

After this `x ? x : fallback` can be checked safely.

**Comment by Rel1cx** [comment]
I'll merge it after you've done the changes.

File: packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts (hunk 2 above)
Comment (hyoban) [reply packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts:232]: It looks like `inspectVariantTypes([testType])` doesn't get the correct type for test case `someCondition?: number`. Because it is a union type, we need to pass `testType.types`.
```
types === ['object']
```

**Review by Rel1cx — APPROVED** [review]
LGTM.
